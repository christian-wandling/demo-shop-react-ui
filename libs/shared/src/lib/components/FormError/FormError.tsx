import { FormErrorProps } from '../../models/formErrorProps';

/**
 * Component that displays form validation error messages.
 *
 * This component renders error messages for form fields that fail validation.
 * It accepts the field name and validation error as props and displays
 * appropriate error messages.
 *
 * @example
 * <FormError
 *  fieldName="Email"
 *  error={errors.email}
 * />
 */
export const FormError = ({ fieldName = 'Field', error }: FormErrorProps) => {
  if (!error) return null;

  return (
    <div className="text-xs text-red-600" role="alert">
      {error.type === 'required' && <div>{fieldName} is required</div>}

      {error.type === 'email' && <div>Email invalid</div>}
      {/* Add more validation type checks as needed */}
    </div>
  );
};
