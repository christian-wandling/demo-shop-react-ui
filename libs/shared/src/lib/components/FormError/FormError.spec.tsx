import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { FormError } from './FormError';
import { FieldError } from 'react-hook-form';

describe('FormError component', () => {
  it('renders nothing when no error is provided', () => {
    const { container } = render(<FormError fieldName="Email" error={undefined} />);
    expect(container.firstChild).toBeNull();
  });

  it('renders required error message with the provided field name', () => {
    const error = { type: 'required' } as FieldError;
    render(<FormError fieldName="Email" error={error} />);

    expect(screen.getByRole('alert')).toBeTruthy();
    expect(screen.getByText('Email is required')).toBeTruthy();
  });

  it('renders email error message for email type errors', () => {
    const error = { type: 'email' } as FieldError;
    render(<FormError fieldName="Email" error={error} />);

    expect(screen.getByRole('alert')).toBeTruthy();
    expect(screen.getByText('Email invalid')).toBeTruthy();
  });

  it('uses default field name when none is provided', () => {
    const error = { type: 'required' } as FieldError;
    render(<FormError error={error} />);

    expect(screen.getByRole('alert')).toBeTruthy();
    expect(screen.getByText('Field is required')).toBeTruthy();
  });

  it('applies the correct CSS classes for styling', () => {
    const error = { type: 'required' } as FieldError;
    render(<FormError fieldName="Email" error={error} />);

    const errorElement = screen.getByRole('alert');
    expect(errorElement.classList).toContain('text-xs');
    expect(errorElement.classList).toContain('text-red-600');
  });

  describe('Snapshots', () => {
    it('matches the snapshot for no errors', () => {
      const { baseElement } = render(<FormError fieldName="Email" error={undefined} />);
      expect(baseElement).toMatchSnapshot();
    });

    it('matches the snapshot for errors', () => {
      const error = { type: 'required' } as FieldError;
      const { baseElement } = render(<FormError fieldName="Email" error={error} />);
      expect(baseElement).toMatchSnapshot();
    });
  });
});
