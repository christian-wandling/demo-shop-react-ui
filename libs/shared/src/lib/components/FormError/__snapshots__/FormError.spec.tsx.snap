// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormError component > Snapshots > matches the snapshot for errors 1`] = `
<body>
  <div>
    <div
      class="text-xs text-red-600"
      role="alert"
    >
      <div>
        Email
         is required
      </div>
    </div>
  </div>
</body>
`;

exports[`FormError component > Snapshots > matches the snapshot for no errors 1`] = `
<body>
  <div />
</body>
`;
